### PR TITLE
QC framework build metrics

### PR DESCRIPTION
**Pull request by jasonwalker80**
Initial revision of a command to dump build QC metrics.

### CODE REVIEW

**Comment by ernfrid** [comment]
Would be better if this had tests of the code itself.

+1

File: lib/perl/Genome/Qc/Command/BuildMetrics.pm [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    my $self = shift;
+
+    my @metrics;
+    for my $build ($self->builds) {
+        my @qc_results = grep {$_->isa('Genome::Qc::Result')} $build->results;
+        for my $qc_result (@qc_results) {
+            my $as = $qc_result->alignment_result;
+            my @instrument_data = $as->instrument_data;
+            if (@instrument_data > 1) {
+                $self->error_message('Please add support for merged alignment results and multiple instrument data QC results!');
```
Comment (apregier) [inline lib/perl/Genome/Qc/Command/BuildMetrics.pm]: You could easily support this by putting the instrument data ids in a comma-separated list (for example) - I think everything else is generic enough that you wouldn't need to change anything.

**Comment by apregier** [comment]
+1 Code looks good to me, although I agree a test will help prevent future problems.  You may need to write a quick Genome::Test::Factory module for a qc result

File: lib/perl/Genome/Qc/Command/BuildMetrics.pm (hunk 1 above)
Comment (jasonwalker80) [inline lib/perl/Genome/Qc/Command/BuildMetrics.pm]: I expect the metric names to be very different for results with multiple instrument data.  The generic separated value writer can handle missing values.  Another option is they become a separate list of `@merged_metrics` with a different writer.

I was waiting to know more once we have a QC config that defines both per read group and merged result QC in the same build.

**Comment by jasonwalker80** [comment]
I started looking at making a qc result test factory and that escalated quickly.  I would need an AlignedBamResult, SoftwareResult and other objects to make it work.  Perhaps I need a nudge on how to make a Genome::Test::Factory::Qc::Result.  There seem to be many dependencies to set one up correctly.

**Comment by jasonwalker80** [comment]
@susannasiebert is helping me through the Genome::Test::Factory::Qc::Result.  There are several changes to other test factories that are needed for this to work.

**Comment by jasonwalker80** [comment]
@ernfrid , @apregier and @susannasiebert I pushed a few commits.  The first one is the necessary changes to make a test factory qc result.  An additional commit improves the AR factory to allow the use of the same build for all results.

**Comment by apregier** [comment]
Would it work to simply call `$test_alignment_result->add_user`
instead of the rigmarole of passing the build around?  Or is it mostly an optimization so that only one test build is created?

File: lib/perl/Genome/Qc/Command/BuildMetrics.t [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+my $pkg = 'Genome::Qc::Command::BuildMetrics';
+use_ok($pkg);
+
+my $cmd = Genome::Qc::Command::BuildMetrics->create(
+   builds => $test_build,
+);
+isa_ok($cmd,'Genome::Qc::Command::BuildMetrics');       
+ok($cmd->execute,'execute build-metrics command');
+
```
Comment (apregier) [inline lib/perl/Genome/Qc/Command/BuildMetrics.t]: If you want to capture the output in a file and compare it, you could use compare_ok with the replace option to fill in the instrument data and build ids, which will change from run to run:
```
use Genome::Utility::Test qw(compare_ok);
my $replace = {
    INSTRUMENT_DATA_ID => $test_inst_data_1->id,
    BUILD_ID => $test_build->id
};
compare_ok($output, $expected_output, replace => $replace, 
    name => "Build metrics printed correctly");
```
Your expected output file would look like:
```
build_id\tinstrument_data_id\tmetric_A\tmetric_B\tmetric_C\tmetric_D
BUILD_ID\tINSTRUMENT_DATA_ID\t1\t2\t3\t4
```

**Comment by jasonwalker80** [comment]
Passing the build was an optimization to avoid the unnecessary creation of additional models and builds. Your solution might work for this test, but it also seemed like a feature other tests might need.

**Comment by jasonwalker80** [comment]
Update to command to add output file option.  Also added test to compare to expected output file.

**Comment by apregier** [comment]
:+1:

File: lib/perl/Genome/Qc/Command/BuildMetrics.t [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+use Test::More;
+use Genome::Utility::Test qw(compare_ok);
+
+use Genome::Test::Factory::InstrumentData::Solexa;
+use Genome::Test::Factory::InstrumentData::AlignmentResult;
+use Genome::Test::Factory::Model::SomaticValidation;
+use Genome::Test::Factory::Build;
+use Genome::Test::Factory::Qc::Result;
+
+use Sub::Install qw (reinstall_sub);
```
Comment (ernfrid) [inline lib/perl/Genome/Qc/Command/BuildMetrics.t]: nitpick - space after qw breaks style from lines above.

File: lib/perl/Genome/Test/Factory/InstrumentData/AlignmentResult.pm [hunk 4]
```diff
@@ -9,9 +9,17 @@ use Genome::Test::Factory::SoftwareResult::User;
...
 sub generate_obj {
     my $self = shift;
+    my %params = @_;
 
-    my $result_users = Genome::Test::Factory::SoftwareResult::User->setup_user_hash();
-    return Genome::InstrumentData::AlignmentResult::Bwa->__define__(@_, _user_data_for_nested_results => $result_users);
+    my $build = delete($params{'build'});
+    
+    my $result_users;
+    unless ($build) {
```
Comment (ernfrid) [inline lib/perl/Genome/Test/Factory/InstrumentData/AlignmentResult.pm]: I find the ```unless ($build)``` construct to be touch difficult to interpret vs a more standard ```if ($build)``` (with proper rearrangement below of course).

**Comment by ernfrid** [comment]
+1 one from me (I think my nitpicks are optional) with the caveat that I'm familiar with the Test::Factory paradigm.

**Pull request merged**